### ISSUE TITLE
DD-PPO does not all reduce gradients

### ISSUE THREAD

**Issue by Purpleslz** [facebookresearch/habitat-api#346]
## 🐛 Bug

DD-PPO does not all reduce gradients during the backward call, because `reducer.prepare_for_backward` is not being called during training process.

The problem is in this line: https://github.com/facebookresearch/habitat-api/blob/v0.1.4/habitat_baselines/rl/ddppo/algo/ddppo.py#L96

```
class DecentralizedDistributedMixin:

    ...
    def before_backward(self, loss):
         # ...
         self.reducer.prepare_for_backward(..)


# Mixin goes second that way the PPO __init__ will still be called
class DDPPO(PPO, DecentralizedDistributedMixin):  
    # Here PPO and Mixin both have "before_backward" method, 
    # DDPPO will call PPO's not the Mixin's.
    pass
```

And here is a quick fix:
```
class DecentralizedDistributedMixin:
    ...


# Mixin goes second that way the PPO __init__ will still be called
class DDPPO(PPO, DecentralizedDistributedMixin):  

      # Move before_backward to DDPPO
      def before_backward(self, loss):
         # ...
         self.reducer.prepare_for_backward(..)
```

**Comment by erikwijmans** [comment]
It's actually easier to fix than that, I don't know what happened, but it should `class DDPPO(DecentralizedDistributedMixin, PPO)`.

**Comment by Xiaoming-Zhao** [comment]
@erikwijmans Hi, may I know whether the main aim of defining `DecentralizedDistributedMixin` class is for manually adding `loss` into the `Reducer` of PyTorch's `DecentralizedDistributedMixin`? Because the agent in DDPPO does not utilize `forward` pass, I think the [automatically processing](https://github.com/pytorch/pytorch/blob/v1.2.0/torch/nn/parallel/distributed.py#L435) here will not work.

**Comment by erikwijmans** [comment]
Yes, that is correct, since the forward pass in not used by the agents, we need to manually make the call to `reducer.prepare_for_backward`.
